File: 1168-duplicate-zeros/1168-duplicate-zeros.py
class Solution:
    def duplicateZeros(self, arr: List[int]) -> None:
        """
        Do not return anything, modify arr in-place instead.
        """
        length = len(arr)
        
        zeroes = 0
        for i in arr:
            if i == 0:
                zeroes += 1
        
        copyFrom = length - 1
        copyTo = length + zeroes - 1

        while copyFrom != copyTo:
            if copyTo < length:
                arr[copyTo] = arr[copyFrom]
            copyTo -= 1
            if arr[copyFrom] == 0:
                if copyTo < length:
                    arr[copyTo] = arr[copyFrom]
                copyTo -= 1
            copyFrom -= 1

File: 1168-duplicate-zeros/1168-duplicate-zeros.java
class Solution {
    public void duplicateZeros(int[] arr) {
        int len = arr.length;
        int zeroes = 0;

        // Count the number of zeroes
        for (int num : arr) {
            if (num == 0) {
                zeroes++;
            }
        }

        int copyFrom = len - 1;
        int copyTo = len + zeroes - 1;

        // Traverse backwards
        while (copyFrom < copyTo) {
            // Only write if within bounds
            if (copyTo < len) {
                arr[copyTo] = arr[copyFrom];
            }

            // If element is 0, write it twice
            if (arr[copyFrom] == 0) {
                copyTo--;
                if (copyTo < len) {
                    arr[copyTo] = 0;
                }
            }

            copyFrom--;
            copyTo--;
        }
    }
}
